doc/tests.rst

```diff
@@ -448,8 +449,160 @@ which displays nicer looking diffs that the standard ``diff``
+
+* Comments - Anything that doesn't start with a 2 space indentation is a comment
+
+* Commands - A command starts with 2 spaces followed by a ``$``. It is executed
+  in the shell and the output is diffed against the output below. In this
+  example, there's no output yet.
+
+To run the test and promote the results:
+
+.. code:: bash
+
+   $ dune runtest
+   $ dune promote
+
+We now see the output of the command:
+
+.. code:: bash
+
+   Simplest possible cram test
+     $ echo "testing"
+     testing
+
+This is the main advantage of expect tests. We don't need to write assertions
+manually, instead we detect failure when the command produces a different output
+than what is recorded in the test script.
+
+For example, here's an example of how we'd test the ``wc`` utility. ``wc.t``:
+.. code:: bash
+
+   We create a fixture called foo
+     $ cat >foo <<EOF
+     > foo
+     > bar
+     > baz
+     > EOF
+
+   After creating the fixture, we want to verify that ``wc`` gives us the right
+   result:
+     $ wc -l foo | awk '{ print $1 }'
+     4
+
+The above example uses the here doc syntax to pipe the subsequent lines to
+``cat``. This is convenient for creating small fixtures.
+
+Directory Tests
+---------------
+
+In the above example we used ``cat`` to create the fixture, but what if there
+are too many fixtures to comfortably fit in test file? Or some of the fixtures
+are binary? It's possible to include the fixtures as normal files or directories
+provided the test is defined as a directory. The name of the test directory must
+end with ``.t`` and must include a ``run.t`` as the test script. Everything else
+in that directory is treated as raw data for the test. It's not possible to
+define rules using ``dune`` files in such a directory.
+
+We convert the ``wc`` test above into a directory test ``wc.t``:
+
+.. code:: bash
+
+   $ ls wc.t
+     run.t foo.txt bar/
+
+This defines a directory test ``wc.t`` which must include a ``run.t`` file as
+the test script, with ``fool.txt`` and ``bar`` are fixtures. We may then access
+their contents in the test script ``run.t``:
+
+.. code:: bash
+
+   $ wc -l foo | awk '{ print $1 }'
+   4
+   $ wc -l $(ls bar) | awk '{ print $1 }'
+   1231
+
+Test Options
+------------
+
+When testing binaries, it's important to to specify a dependency on the binary
+for two reasons:
+
+- Dune must know to re-run the test when a dependency changes
+
+- The dependencies must be specified to guarantee that they are visible to the
+  test when running it.
+
+We can specify dependencies using the ``deps`` field using the usual syntax:
+
+.. code:: bash
+
+   (cram
+    (deps ../foo.exe))
+
+This introduces a dependency on ``foo.exe`` on all cram tests in this directory.
+To apply the stanza to a particular test, it's possible to use ``applies_to``
+field:
+
+.. code:: bash
+
+   (cram
+    (applies_to * \ foo bar)
+    (deps ../foo.exe))
+
+We use the predicate language to apply this stanza to all tests in this
```

Comment:
Do we have a section describing the predicate language? If yes, we should reference it here